Speculation Rules: implement Navigation Timing's deliveryType

yoavweiss opened the pull request:
#### f009b61f8d117e884ec2750b47bf387d85695d77
<pre>
Speculation Rules: implement Navigation Timing&apos;s deliveryType
<a href="https://bugs.webkit.org/show_bug.cgi?id=300586">https://bugs.webkit.org/show_bug.cgi?id=300586</a>

Reviewed by Alex Christensen.

This PR exposes the `deliveryType` attribute on Resource Timing, with an empty string value in most cases.
It then sets that value to &quot;navigational-prefetch&quot; for prefetched documents, exposed in their navigation timing entries.

No new tests, but existing test expectations progress.

* LayoutTests/imported/w3c/web-platform-tests/resource-timing/idlharness.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/resource-timing/idlharness.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/navigation-timing-delivery-type.https_prefetch=false&amp;bypass_cache=false-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/navigation-timing-delivery-type.https_prefetch=false&amp;bypass_cache=true-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/navigation-timing-delivery-type.https_prefetch=true&amp;bypass_cache=false-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/navigation-timing-delivery-type.https_prefetch=true&amp;bypass_cache=true-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/prefetch-uses-cache.sub.https_cross-site-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/prefetch-uses-cache.sub.https_same-site-expected.txt:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::notifyFinished): Set the `fromPrefetch` parameter on metrics that come from a prefetched resource.
(WebCore::DocumentLoader::commitData): Set the `fromPrefetch` parameter on metrics where there&apos;s an in-flight prefetch with the same URL.
* Source/WebCore/loader/DocumentPrefetcher.cpp:
(WebCore::DocumentPrefetcher::wasPrefetched const): Returns true if there&apos;s an in-flight prefetch with that URL.
(WebCore::DocumentPrefetcher::takePrefetchedNetworkLoadMetrics):
* Source/WebCore/loader/DocumentPrefetcher.h:
* Source/WebCore/loader/ResourceTiming.h:
(WebCore::ResourceTiming::deliveryType const): Getter.
* Source/WebCore/page/PerformanceResourceTiming.h:
(WebCore::PerformanceResourceTiming::deliveryType const):
(WebCore::PerformanceResourceTiming::resourceTiming):
* Source/WebCore/page/PerformanceResourceTiming.idl: Add the deliveryType attribute.
* Source/WebCore/platform/network/NetworkLoadMetrics.cpp:
(WebCore::NetworkLoadMetrics::NetworkLoadMetrics): Initialize fromPrefetch.
(WebCore::NetworkLoadMetrics::isolatedCopy const): Copy fromPrefetch.
* Source/WebCore/platform/network/NetworkLoadMetrics.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/303527@main">https://commits.webkit.org/303527@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/50c492c7d663945c41b2680c8283c9c8e45e9431

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132692 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5187 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43785 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140220 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84719 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134562 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5435 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4946 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101449 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68748 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135638 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118890 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82242 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1451 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83459 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/112671 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37005 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142875 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4857 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37593 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109829 "Found 1 new test failure: fast/scrolling/rtl-scrollbars-sticky-document.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4939 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4203 "Found 1 new test failure: http/tests/ipc/ipc-fetch-task-message-crash.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110006 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27887 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3706 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115161 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58339 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4911 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33508 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4749 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68362 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5002 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4868 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/loader/ResourceTiming.h:
```diff
@@ -76,6 +79,7 @@ class ResourceTiming {
     ResourceLoadTiming m_resourceLoadTiming;
     NetworkLoadMetrics m_networkLoadMetrics;
     Vector<ServerTiming> m_serverTiming;
+    String m_deliveryType;
```
It seems excessive to store a string when it is one of two states.  Can't we just store a bool and return empty or "navigational-prefetch" from ResourceTiming::deliveryType()?

yoavweiss replied:
This could eventually be a tri-state: "navigational-prefetch", "cache" or the empty string. (or more as this attribute evolves)

But given we don't support the "cache" signal yet (nor its `transferSize` proxy), I'm happy to change it to a bool and then modify to an enum or a string if needed in the future.

yoavweiss replied:
Done

achristensen07 commented on Source/WebCore/page/Performance.cpp:
```diff
@@ -396,6 +396,9 @@ void Performance::enqueueLargestContentfulPaint(Ref<LargestContentfulPaint>&& pa
     m_navigationTiming = PerformanceNavigationTiming::create(m_timeOrigin, resource, timing, metrics, document.eventTiming(), document.securityOrigin(), documentLoader.triggeringAction().type());
+
+    if (metrics.fromPrefetch)
+        m_navigationTiming->resourceTiming().setIsNavigationalPrefetch();
```
We pass metrics to the PerformanceNavigationTiming constructor.  It's generally an anti-pattern to pass a parameter to a constructor then immediately use it to call a setter.

yoavweiss replied:
Removed

achristensen07 commented on Source/WebCore/loader/ResourceTiming.h:
```diff
@@ -76,6 +79,7 @@ class ResourceTiming {
     ResourceLoadTiming m_resourceLoadTiming;
     NetworkLoadMetrics m_networkLoadMetrics;
     Vector<ServerTiming> m_serverTiming;
+    bool m_isNavigationalPrefetch { false };
```
This is redundant information.  We already have m_networkLoadMetrics.

yoavweiss replied:
Oops, indeed! removed

achristensen07 commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -1337,7 +1339,10 @@ void DocumentLoader::commitData(const SharedBuffer& data)
-                window->protectedPerformance()->addNavigationTiming(*this, document, *m_mainResource, timing(), metrics ? *metrics : NetworkLoadMetrics::emptyMetrics());
+                NetworkLoadMetrics finalMetrics = metrics ? *metrics : NetworkLoadMetrics::emptyMetrics();
+                if (RefPtr frameLoader = this->frameLoader())
+                    finalMetrics.fromPrefetch = frameLoader->documentPrefetcher().wasPrefetched(url());
```
This feels a little shaky.  It's strange that takePrefetchedNetworkLoadMetrics takes the metrics but leaves the CachedResourceHandle in m_prefetchedData.  This design assumes m_prefetchedData is never cleaned up, which isn't great.  m_prefetchedData is added to but never removed from.

yoavweiss replied:
Removed the handle as part of takePrefetchedNetworkLoadMetrics

achristensen07 commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -7363,6 +7363,7 @@ class WebCore::NetworkLoadMetrics {
     bool reusedConnection();
     bool doesFailTAOCheck();
     bool crossOriginRedirect();
+    bool fromPrefetch;
```
It's a bit inconsistent to use the accessor methods for the others but access the member directly for this one.  The simplest way to make it consistent would be to add an isFromPrefetch() const accessor, but maybe we could just use [BitField] to encode all the members directly.

yoavweiss replied:
Moved to [BitField]

yoavweiss committed:
Speculation Rules: implement Navigation Timing's deliveryType
https://bugs.webkit.org/show_bug.cgi?id=300586

Reviewed by Alex Christensen.

This PR exposes the `deliveryType` attribute on Resource Timing, with an empty string value in most cases.
It then sets that value to "navigational-prefetch" for prefetched documents, exposed in their navigation timing entries.

No new tests, but existing test expectations progress.

* LayoutTests/imported/w3c/web-platform-tests/resource-timing/idlharness.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/resource-timing/idlharness.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/navigation-timing-delivery-type.https_prefetch=false&bypass_cache=false-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/navigation-timing-delivery-type.https_prefetch=false&bypass_cache=true-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/navigation-timing-delivery-type.https_prefetch=true&bypass_cache=false-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/navigation-timing-delivery-type.https_prefetch=true&bypass_cache=true-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/prefetch-uses-cache.sub.https_cross-site-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/prefetch-uses-cache.sub.https_same-site-expected.txt:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::notifyFinished): Set the `fromPrefetch` parameter on metrics that come from a prefetched resource.
(WebCore::DocumentLoader::commitData): Set the `fromPrefetch` parameter on metrics where there's an in-flight prefetch with the same URL.
* Source/WebCore/loader/DocumentPrefetcher.cpp:
(WebCore::DocumentPrefetcher::wasPrefetched const): Returns true if there's an in-flight prefetch with that URL.
(WebCore::DocumentPrefetcher::takePrefetchedNetworkLoadMetrics):
* Source/WebCore/loader/DocumentPrefetcher.h:
* Source/WebCore/loader/ResourceTiming.h:
(WebCore::ResourceTiming::deliveryType const): Getter.
* Source/WebCore/page/PerformanceResourceTiming.h:
(WebCore::PerformanceResourceTiming::deliveryType const):
(WebCore::PerformanceResourceTiming::resourceTiming):
* Source/WebCore/page/PerformanceResourceTiming.idl: Add the deliveryType attribute.
* Source/WebCore/platform/network/NetworkLoadMetrics.cpp:
(WebCore::NetworkLoadMetrics::NetworkLoadMetrics): Initialize fromPrefetch.
(WebCore::NetworkLoadMetrics::isolatedCopy const): Copy fromPrefetch.
* Source/WebCore/platform/network/NetworkLoadMetrics.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/303527@main
